x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/edit_policy.tsx

```diff
@@ -188,215 +236,217 @@ export const EditPolicy: React.FunctionComponent<Props> = ({
-                          id="xpack.indexLifecycleMgmt.editPolicy.saveButton"
-                          defaultMessage="Save policy"
+                          id="xpack.indexLifecycleMgmt.editPolicy.cancelButton"
+                          defaultMessage="Cancel"
                         />
-                      )}
-                    </EuiButton>
-                  </EuiFlexItem>
-
-                  <EuiFlexItem grow={false}>
-                    <EuiButtonEmpty data-test-subj="cancelTestPolicy" onClick={backToPolicyList}>
+                      </EuiButtonEmpty>
+                    </EuiFlexItem>
+                  </EuiFlexGroup>
+                </EuiFlexItem>
+
+                <EuiFlexItem grow={false}>
+                  <EuiButtonEmpty onClick={togglePolicyJsonFlyout} data-test-subj="requestButton">
+                    {isShowingPolicyJsonFlyout ? (
                       <FormattedMessage
-                        id="xpack.indexLifecycleMgmt.editPolicy.cancelButton"
-                        defaultMessage="Cancel"
+                        id="xpack.indexLifecycleMgmt.editPolicy.hidePolicyJsonButto"
+                        defaultMessage="Hide request"
                       />
-                    </EuiButtonEmpty>
-                  </EuiFlexItem>
-                </EuiFlexGroup>
-              </EuiFlexItem>
-
-              <EuiFlexItem grow={false}>
-                <EuiButtonEmpty onClick={togglePolicyJsonFlyout} data-test-subj="requestButton">
-                  {isShowingPolicyJsonFlyout ? (
-                    <FormattedMessage
-                      id="xpack.indexLifecycleMgmt.editPolicy.hidePolicyJsonButto"
-                      defaultMessage="Hide request"
-                    />
-                  ) : (
-                    <FormattedMessage
-                      id="xpack.indexLifecycleMgmt.editPolicy.showPolicyJsonButto"
-                      defaultMessage="Show request"
-                    />
-                  )}
-                </EuiButtonEmpty>
-              </EuiFlexItem>
-            </EuiFlexGroup>
-
-            {isShowingPolicyJsonFlyout ? (
-              <PolicyJsonFlyout
-                policyName={policy.name || ''}
-                existingPolicy={existingPolicy}
-                policy={policy}
-                close={() => setIsShowingPolicyJsonFlyout(false)}
-              />
-            ) : null}
+                    ) : (
+                      <FormattedMessage
+                        id="xpack.indexLifecycleMgmt.editPolicy.showPolicyJsonButto"
+                        defaultMessage="Show request"
+                      />
+                    )}
+                  </EuiButtonEmpty>
+                </EuiFlexItem>
+              </EuiFlexGroup>
+
+              {isShowingPolicyJsonFlyout ? (
+                <PolicyJsonFlyout
+                  policyName={policy.name || ''}
+                  readPolicy={() => {
+                    const serializedPolicy = form.getFormData();
```

Comment:
Here it would be good to check if the form is valid because `getFormData` calls the serializer and the form lib guarantees to the serializers that the values are valid (so each serializer does not need to check the field validity again).

I could enforce this by returning null in the form lib in case the form is invalid, but that would create a breaking change.

This implies that if the form is not valid, we should not merge the data as it could potentially throw.

We could approach it differently:

```js
// Have a ref with the serialized form
const serializedPolicy = useRef(null);
const [isShowingPolicyJsonFlyout, setIsShowingPolicyJsonFlyout] = useState(false);
const { validate: validateForm, getFormData, isValid: isFormValid } = form;

useEffect(() => {
  if (isShowingPolicyJsonFlyout) {
    validateForm().then((isValid) => {
      if (isValid) {
        // We are sure that serializedPolicy is valid
        serializedPolicy.current = getFormData();
      }
    });
  } else {
    serializedPolicy.current = null;
  }
}, [isShowingPolicyJsonFlyout, validateForm, getFormData]);

// You can then use the "isFormValid" state below to only render JSON flyout if it is "true"
```

Another approach could be to use the `await form.submit()` when clicking the show request button. If there are error, ask the user to first correct them. WDYT?

Reply:
@jloleysens UX-wise, it might make sense to always allow the flyout to open, but to show the a danger callout inside the flyout that says something like, "Fix the errors in the policy to see its JSON". If I click a button and it doesn't do what I expect, I'm not sure I'd find it natural to then look for the form's validation state to figure out what went wrong.

Reply:
Perhaps we can just disable the show flyout button if the form state is invalid 😅? We do the same with the submit button?

FWIW I kept the flow the same as the current ILM policy UX - which will show the form JSON even if it is invalid. Happy to change this behaviour now.